examples/tutorials/device_asr.py

```diff
@@ -0,0 +1,330 @@
+#
+# In the above case, the following value can be used to stream from microphone.
+#
+# .. code::
+#
+#    Streamer(
+#        src = "audio=@device_cm_{33D9A762-90C8-11D0-BD43-00A0C911CE86}\wave_{BF2B8AE1-10B8-4CA4-A0DC-D02E18A56177}",
+#        format = "dshow",
+#    )
+#
+
+######################################################################
+# 3. Data acquisition
+# -------------------
+#
+# Streaming audio from microphone requires to properly time acquiring
+# data form hardware. Failing to do so indtorduces discontinuity in
+# data stream.
+#
+# For this reason, we will run the data acquisition in a subprocess.
+#
+# Firstly, we create a helper function that encupsulates the whole
+# process executed in the subprocess.
+#
+# This function initializes the streaming API, acquire data then
+# put it in a queue, which the main process is watching.
+#
+
+import torch
+import torchaudio
+
+
+# The data acquisition process will stop after this number of steps.
+# This eliminates the need of process synchronization and makes this
+# tutorial simple.
+NUM_ITER = 100
+
+
+def stream(q, format, src, segment_length, sample_rate):
+    from torchaudio.prototype.io import Streamer
+
+    print("Building Streamer...")
+    streamer = Streamer(src, format=format)
+    streamer.add_basic_audio_stream(frames_per_chunk=segment_length, sample_rate=sample_rate)
+
+    print(streamer.get_src_stream_info(0))
+    print(streamer.get_out_stream_info(0))
+
+    print("Streaming...")
+    print()
+    stream_iterator = streamer.stream(timeout=-1, backoff=1.0)
+    for _ in range(NUM_ITER):
+        (chunk,) = next(stream_iterator)
+        q.put(chunk)
+
+
+######################################################################
+#
+# The notable difference from the non-device streaming is that,
+# we provide ``timeout`` and ``backoff`` parameters to ``stream`` method.
+#
+# When acquiring data, if the rate of acquisition request is faster
+# than what hardware can prepare the data, then the underlying implementation
+# reports special error code, and expects client code to retry.
+#
+# Precise timing is the key for smooth streaming. Reporting this error
+# from low level implementation to all the way back to Python layer,
+# before retrying adds undesired overhead.
+# For this reason, the retry behavior is implemented in C++ layer, and
+# ``timeout`` and ``backoff`` parameters allow client code to control the
+# behavior.
+#
+# For the detail of ``timeout`` and ``backoff`` parameters, please refer
+# to the documentation of
+# :py:meth:`~torchaudio.prototype.io.Streamer.stream` method.
+#
+# .. note::
+#
+#    The proper value of ``backoff`` depends on the system configuration.
+#    One way to see if ``backoff`` value is appropriate is to save the
+#    series of acquired chunks as a continuous audio and listen to it.
+#    If ``backoff`` value is too large, then the data stream is discontinuous.
+#    The resulting audio sounds sped up.
```

Comment:
is this because the streamer buffer wouldn't be flushed quickly enough due to the excessively large backoff? so buffer contents end up getting overwritten, and therefore sections of the data stream end up going missing?

Reply:
Yes, the hardware buffer will be refilled at a rate faster than the torchaudio code acquiring the data.
The data client code receive miss portions, which would sound like skipped, thus sped up.